File: android/app/src/main/java/dev/framework/example/FrameworkRuntimePlugin.java
package dev.framework.example;

import android.content.res.AssetManager;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.webkit.MimeTypeMap;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.time.Instant;
import java.util.Locale;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "FrameworkRuntime")
public class FrameworkRuntimePlugin extends Plugin {

    private static final String ASSET_ROOT = "public/app-assets";
    private RuntimeDatabase database;

    @Override
    public void load() {
        database = new RuntimeDatabase(getContext());
    }

    @PluginMethod
    public void invoke(PluginCall call) {
        JSObject request = call.getObject("request");
        if (request == null) {
            call.reject("Missing runtime request");
            return;
        }

        try {
            JSObject response = new JSObject();
            response.put("result", handle(request));
            call.resolve(response);
        } catch (RuntimeAssetException error) {
            JSObject response = new JSObject();
            response.put("result", runtimeFailure(error));
            call.resolve(response);
        } catch (Exception error) {
            call.reject(error.getMessage(), error);
        }
    }

    private Object handle(JSObject request) throws JSONException, RuntimeAssetException {
        String tag = request.getString("_tag");
        if (tag == null) {
            throw new JSONException("Missing runtime request tag");
        }

        switch (tag) {
            case "Health":
                return new JSObject().put("status", "ok").put("runtime", "mobile:android");
            case "GetUserSetting":
                return getUserSetting(requiredString(request, "key"));
            case "SetUserSetting":
                return setUserSetting(requiredString(request, "key"), requiredString(request, "value"));
            case "ListAppStorageEntries":
                return listAppStorageEntries(requiredString(request, "appId"), requiredString(request, "collection"));
            case "AppendAppStorageEntry":
                return appendAppStorageEntry(
                    requiredString(request, "appId"),
                    requiredString(request, "collection"),
                    requiredString(request, "value")
                );
            case "ResolveAppAsset":
                return resolveAppAsset(requiredString(request, "appId"), requiredString(request, "path"));
            default:
                throw new JSONException("Unknown runtime request tag: " + tag);
        }
    }

    private JSObject getUserSetting(String key) {
        SQLiteDatabase db = database.getReadableDatabase();
        try (
            Cursor cursor = db.rawQuery("SELECT key, value, updated_at FROM user_settings WHERE key = ?", new String[] { key })
        ) {
            if (!cursor.moveToFirst()) {
                return null;
            }
            return userSetting(cursor.getString(0), cursor.getString(1), cursor.getString(2));
        }
    }

    private JSObject setUserSetting(String key, String value) {
        String updatedAt = Instant.now().toString();
        SQLiteDatabase db = database.getWritableDatabase();
        db.execSQL(
            "INSERT INTO user_settings (key, value, updated_at) VALUES (?, ?, ?) " +
            "ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at",
            new Object[] { key, value, updatedAt }
        );
        return userSetting(key, value, updatedAt);
    }

    private JSArray listAppStorageEntries(String appId, String collection) throws JSONException {
        SQLiteDatabase db = database.getReadableDatabase();
        JSArray entries = new JSArray();
        try (
            Cursor cursor = db.rawQuery(
                "SELECT id, app_id, collection, value, created_at FROM app_storage_entries " +
                "WHERE app_id = ? AND collection = ? ORDER BY id ASC",
                new String[] { appId, collection }
            )
        ) {
            while (cursor.moveToNext()) {
                entries.put(appStorageEntry(cursor));
            }
        }
        return entries;
    }

    private JSObject appendAppStorageEntry(String appId, String collection, String value) {
        String createdAt = Instant.now().toString();
        SQLiteDatabase db = database.getWritableDatabase();
        db.execSQL(
            "INSERT INTO app_storage_entries (app_id, collection, value, created_at) VALUES (?, ?, ?, ?)",
            new Object[] { appId, collection, value, createdAt }
        );
        long id = lastInsertRowId(db);
        return appStorageEntry(id, appId, collection, value, createdAt);
    }

    private JSObject resolveAppAsset(String appId, String rawPath) throws RuntimeAssetException {
        String normalizedPath = normalizeAssetPath(rawPath);
        String assetPath = ASSET_ROOT + "/" + appId + "/" + normalizedPath;
        AssetManager assets = getContext().getAssets();

        try (InputStream stream = assets.open(assetPath)) {
            long size = streamSize(stream);
            return new JSObject()
                .put("contentType", contentType(normalizedPath))
                .put("appId", appId)
                .put("path", normalizedPath)
                .put("size", size)
                .put("version", "asset-" + size);
        } catch (IOException error) {
            if (!assetExists(assets, ASSET_ROOT + "/" + appId)) {
                throw new RuntimeAssetException(appId, rawPath, "UnknownApp", "Unknown app assets: " + appId);
            }
            throw new RuntimeAssetException(appId, rawPath, "NotFound", "Asset not found: " + rawPath);
        }
    }

    private static JSObject userSetting(String key, String value, String updatedAt) {
        return new JSObject().put("key", key).put("value", value).put("updatedAt", updatedAt);
    }

    private static JSObject appStorageEntry(Cursor cursor) {
        return appStorageEntry(cursor.getLong(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4));
    }

    private static JSObject appStorageEntry(long id, String appId, String collection, String value, String createdAt) {
        return new JSObject()
            .put("id", id)
            .put("appId", appId)
            .put("collection", collection)
            .put("value", value)
            .put("createdAt", createdAt);
    }

    private static JSObject runtimeFailure(RuntimeAssetException error) {
        return new JSObject().put("_tag", "RuntimeMobileFailure").put("error", error.toJson());
    }

    private static String requiredString(JSObject object, String key) throws JSONException {
        String value = object.getString(key);
        if (value == null || value.isEmpty()) {
            throw new JSONException("Missing required string: " + key);
        }
        return value;
    }

    private static String normalizeAssetPath(String path) throws RuntimeAssetException {
        String normalized = path.startsWith("://assets/") ? path.substring("://assets/".length()) : path;
        normalized = normalized.startsWith("/") ? normalized.substring(1) : normalized;
        if (normalized.isEmpty() || normalized.contains("\0")) {
            throw new RuntimeAssetException("", path, "InvalidPath", "Invalid asset path: " + path);
        }
        for (String part : normalized.split("/")) {
            if (part.isEmpty() || part.equals(".") || part.equals("..")) {
                throw new RuntimeAssetException("", path, "InvalidPath", "Invalid asset path: " + path);
            }
        }
        return normalized;
    }

    private static String contentType(String path) {
        int index = path.lastIndexOf(".");
        String extension = index >= 0 ? path.substring(index + 1).toLowerCase(Locale.ROOT) : "";
        switch (extension) {
            case "glb":
                return "model/gltf-binary";
            case "gltf":
                return "model/gltf+json";
            case "gdshader":
                return "text/plain";
            default:
                String mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
                return mime == null ? "application/octet-stream" : mime;
        }
    }

    private static boolean assetExists(AssetManager assets, String path) {
        try {
            String[] children = assets.list(path);
            return children != null;
        } catch (IOException error) {
            return false;
        }
    }

    private static long streamSize(InputStream stream) throws IOException {
        byte[] buffer = new byte[64 * 1024];
        long size = 0;
        int read;
        while ((read = stream.read(buffer)) != -1) {
            size += read;
        }
        return size;
    }

    private static long lastInsertRowId(SQLiteDatabase db) {
        try (Cursor cursor = db.rawQuery("SELECT last_insert_rowid()", null)) {
            cursor.moveToFirst();
            return cursor.getLong(0);
        }
    }

    private static final class RuntimeDatabase extends SQLiteOpenHelper {

        RuntimeDatabase(android.content.Context context) {
            super(context, new File(context.getFilesDir(), "framework.sqlite").getAbsolutePath(), null, 1);
        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            db.execSQL("CREATE TABLE IF NOT EXISTS user_settings (key TEXT PRIMARY KEY, value TEXT NOT NULL, updated_at TEXT NOT NULL)");
            db.execSQL(
                "CREATE TABLE IF NOT EXISTS app_storage_entries (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT, " +
                "app_id TEXT NOT NULL, " +
                "collection TEXT NOT NULL, " +
                "value TEXT NOT NULL, " +
                "created_at TEXT NOT NULL)"
            );
            db.execSQL(
                "CREATE INDEX IF NOT EXISTS app_storage_entries_lookup ON app_storage_entries (app_id, collection, id)"
            );
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
            onCreate(db);
        }
    }

    private static final class RuntimeAssetException extends Exception {

        private final String appId;
        private final String path;
        private final String reason;

        RuntimeAssetException(String appId, String path, String reason, String message) {
            super(message);
            this.appId = appId;
            this.path = path;
            this.reason = reason;
        }

        JSONObject toJson() {
            return new JSObject()
                .put("appId", appId)
                .put("message", getMessage())
                .put("path", path)
                .put("reason", reason);
        }
    }
}

File: ios/App/App/FrameworkRuntimePlugin.swift
import Capacitor
import Foundation

@objc(FrameworkRuntimePlugin)
public class FrameworkRuntimePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FrameworkRuntimePlugin"
    public let jsName = "FrameworkRuntime"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "invoke", returnType: CAPPluginReturnPromise)
    ]

    private let assetRoot = "public/app-assets"
    private var store = RuntimeStore()

    public override func load() {
        store.load()
    }

    @objc func invoke(_ call: CAPPluginCall) {
        guard let request = call.getObject("request") else {
            call.reject("Missing runtime request")
            return
        }

        do {
            call.resolve(["result": try handle(request)])
        } catch let error as RuntimeAssetError {
            call.resolve(["result": runtimeFailure(error)])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    private func handle(_ request: JSObject) throws -> Any {
        guard let tag = request.getString("_tag") else {
            throw RuntimeError.message("Missing runtime request tag")
        }

        switch tag {
        case "Health":
            return ["status": "ok", "runtime": "mobile:ios"]
        case "GetUserSetting":
            return store.userSettings[try requiredString(request, "key")]?.json ?? NSNull()
        case "SetUserSetting":
            let setting = RuntimeStore.UserSetting(
                key: try requiredString(request, "key"),
                value: try requiredString(request, "value"),
                updatedAt: isoNow()
            )
            store.userSettings[setting.key] = setting
            store.save()
            return setting.json
        case "ListAppStorageEntries":
            let appId = try requiredString(request, "appId")
            let collection = try requiredString(request, "collection")
            return store.entries
                .filter { $0.appId == appId && $0.collection == collection }
                .sorted { $0.id < $1.id }
                .map(\.json)
        case "AppendAppStorageEntry":
            let entry = RuntimeStore.AppStorageEntry(
                id: store.nextEntryId,
                appId: try requiredString(request, "appId"),
                collection: try requiredString(request, "collection"),
                value: try requiredString(request, "value"),
                createdAt: isoNow()
            )
            store.nextEntryId += 1
            store.entries.append(entry)
            store.save()
            return entry.json
        case "ResolveAppAsset":
            return try resolveAppAsset(
                appId: try requiredString(request, "appId"),
                rawPath: try requiredString(request, "path")
            )
        default:
            throw RuntimeError.message("Unknown runtime request tag: \(tag)")
        }
    }

    private func resolveAppAsset(appId: String, rawPath: String) throws -> JSObject {
        let normalizedPath = try normalizeAssetPath(rawPath)
        let relativePath = "\(assetRoot)/\(appId)/\(normalizedPath)"
        guard let url = Bundle.main.url(forResource: relativePath, withExtension: nil) else {
            if Bundle.main.url(forResource: "\(assetRoot)/\(appId)", withExtension: nil) == nil {
                throw RuntimeAssetError(
                    appId: appId,
                    path: rawPath,
                    reason: "UnknownApp",
                    message: "Unknown app assets: \(appId)"
                )
            }
            throw RuntimeAssetError(appId: appId, path: rawPath, reason: "NotFound", message: "Asset not found: \(rawPath)")
        }

        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        let size = attributes[.size] as? NSNumber ?? 0
        return [
            "contentType": contentType(normalizedPath),
            "appId": appId,
            "path": normalizedPath,
            "size": size.intValue,
            "version": "asset-\(size.intValue)"
        ]
    }

    private func runtimeFailure(_ error: RuntimeAssetError) -> JSObject {
        [
            "_tag": "RuntimeMobileFailure",
            "error": [
                "appId": error.appId,
                "message": error.message,
                "path": error.path,
                "reason": error.reason
            ]
        ]
    }

    private func requiredString(_ object: JSObject, _ key: String) throws -> String {
        guard let value = object.getString(key), !value.isEmpty else {
            throw RuntimeError.message("Missing required string: \(key)")
        }
        return value
    }

    private func normalizeAssetPath(_ path: String) throws -> String {
        var normalized = path.hasPrefix("://assets/") ? String(path.dropFirst("://assets/".count)) : path
        if normalized.hasPrefix("/") {
            normalized.removeFirst()
        }
        if normalized.isEmpty || normalized.contains("\0") {
            throw RuntimeAssetError(appId: "", path: path, reason: "InvalidPath", message: "Invalid asset path: \(path)")
        }
        for part in normalized.split(separator: "/") {
            if part.isEmpty || part == "." || part == ".." {
                throw RuntimeAssetError(appId: "", path: path, reason: "InvalidPath", message: "Invalid asset path: \(path)")
            }
        }
        return normalized
    }

    private func contentType(_ path: String) -> String {
        switch URL(fileURLWithPath: path).pathExtension.lowercased() {
        case "glb":
            return "model/gltf-binary"
        case "gltf":
            return "model/gltf+json"
        case "gdshader":
            return "text/plain"
        case "jpg", "jpeg":
            return "image/jpeg"
        case "png":
            return "image/png"
        case "webp":
            return "image/webp"
        case "json":
            return "application/json"
        default:
            return "application/octet-stream"
        }
    }

    private func isoNow() -> String {
        ISO8601DateFormatter().string(from: Date())
    }
}

private enum RuntimeError: Error {
    case message(String)
}

private struct RuntimeAssetError: Error {
    let appId: String
    let path: String
    let reason: String
    let message: String
}

private struct RuntimeStore: Codable {
    struct UserSetting: Codable {
        let key: String
        let value: String
        let updatedAt: String

        var json: JSObject {
            ["key": key, "value": value, "updatedAt": updatedAt]
        }
    }

    struct AppStorageEntry: Codable {
        let id: Int
        let appId: String
        let collection: String
        let value: String
        let createdAt: String

        var json: JSObject {
            [
                "id": id,
                "appId": appId,
                "collection": collection,
                "value": value,
                "createdAt": createdAt
            ]
        }
    }

    var userSettings: [String: UserSetting] = [:]
    var entries: [AppStorageEntry] = []
    var nextEntryId = 1

    private var url: URL {
        let directory = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("FrameworkRuntime", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory.appendingPathComponent("runtime-store.json")
    }

    mutating func load() {
        guard let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode(RuntimeStore.self, from: data)
        else {
            return
        }
        self = decoded
    }

    func save() {
        guard let data = try? JSONEncoder().encode(self) else {
            return
        }
        try? data.write(to: url, options: [.atomic])
    }
}
